pyteal/ast/router.py

```diff
@@ -0,0 +1,567 @@
+        If an Application is called with greater than zero Application call arguments  (NOT a bare Application call),
+        the Application MUST always treat the first argument as a method selector and invoke the specified method,
+        regardless of the OnCompletion action of the Application call.
+        This applies to Application creation transactions as well, where the supplied Application ID is 0.
+
+    The `CallConfigs` implementation generalized contract method call such that method call is allowed
+    for certain OnCompletions.
+
+    The `arc4_compliant` method constructs a `CallConfigs` that allows a method call to be executed
+    under any OnCompletion, which is "arc4-compliant".
+    """
+
+    no_op: CallConfig = field(kw_only=True, default=CallConfig.CALL)
+    opt_in: CallConfig = field(kw_only=True, default=CallConfig.NEVER)
+    close_out: CallConfig = field(kw_only=True, default=CallConfig.NEVER)
+    clear_state: CallConfig = field(kw_only=True, default=CallConfig.NEVER)
+    update_application: CallConfig = field(kw_only=True, default=CallConfig.NEVER)
+    delete_application: CallConfig = field(kw_only=True, default=CallConfig.NEVER)
+
+    def is_never(self) -> bool:
+        return all(map(lambda cc: cc == CallConfig.NEVER, astuple(self)))
+
+    @classmethod
+    def arc4_compliant(cls):
+        return cls(
+            no_op=CallConfig.ALL,
+            opt_in=CallConfig.ALL,
+            close_out=CallConfig.ALL,
+            clear_state=CallConfig.ALL,
+            update_application=CallConfig.ALL,
+            delete_application=CallConfig.ALL,
+        )
+
+    def is_arc4_compliant(self) -> bool:
+        return self == self.arc4_compliant()
+
+    def _oc_under_call_config(self, call_config: CallConfig) -> list[EnumInt]:
+        if not isinstance(call_config, CallConfig):
+            raise TealInputError(
+                "generate condition based on CallConfigs should be based on CallConfig"
+            )
+        config_oc_pairs: list[tuple[CallConfig, EnumInt]] = [
+            (self.no_op, OnComplete.NoOp),
+            (self.opt_in, OnComplete.OptIn),
+            (self.close_out, OnComplete.CloseOut),
+            (self.clear_state, OnComplete.ClearState),
+            (self.update_application, OnComplete.UpdateApplication),
+            (self.delete_application, OnComplete.DeleteApplication),
+        ]
+        return [
+            oc
+            for oc_config, oc in config_oc_pairs
+            if (oc_config & call_config) != CallConfig.NEVER
+        ]
+
+    def _has_clear_state(self, cc: CallConfig) -> bool:
+        return (self.clear_state & cc) != CallConfig.NEVER
+
+    def _approval_program_oc(self, cc: CallConfig) -> list[EnumInt]:
+        return list(
+            filter(
+                lambda x: str(x) != str(OnComplete.ClearState),
+                self._oc_under_call_config(cc),
+            )
+        )
+
+
+@dataclass(frozen=True)
+class OnCompleteAction:
```

Comment:
The observation I was making was that it's never possible for `on_create` and `on_call` to have different values that are both non-None. So to simplify, you can store just one value and a `CallConfig` to keep track of whether creation/normal call is supported for it.

Reply:
since it is a `dataclass`, and `__init__` is not overriden, we can still construct instances like `OnCompleteAction(on_create = ..., on_call = ...)`, though it is not straightforward with the factory staticmethod.

Reply:
Yes but there is no good reason to have separate expressions for `on_create` and `on_call` -- we should not allow this.

I'm suggesting this class be simplified to something like:

```python
@dataclass(frozen=True)
class OnCompleteAction:
    handler: Expr | SubroutineFnWrapper | ABIReturnSubroutine | None = field(kw_only=True)
    config: CallConfig = field(kw_only=True)

    def is_empty(self) -> bool:
        return self.config == CallConfig.NEVER
```

Reply:
good point, I changed accordingly on `OnCompleteAction`.